controller/src/main/java/io/pravega/controller/store/PravegaTablesScope.java

```diff
@@ -223,8 +305,80 @@ public void refresh() {
+                                                        .map(key -> storeHelper.getAndUpdateValues(table, key,
+                                                                                                   e -> appendStreamToEntry(stream, e),
+                                                                                                   e -> false, // no need of cleanup
+                                                                                                   context.getRequestId()))
+                                                        .collect(Collectors.toList())));
+    }
+
+    private TableSegmentEntry appendStreamToEntry(String appendValue, TableSegmentEntry entry) {
+        String key = entry.getKey().getKey().toString(StandardCharsets.UTF_8);
+        byte[] array = storeHelper.getArray(entry.getValue());
+        byte[] updatedBytes;
+        if (array.length == 0) {
+            updatedBytes = TagRecord.builder().tagName(key).stream(appendValue).build().toBytes();
+        } else {
+            TagRecord record = TagRecord.fromBytes(array);
+            updatedBytes = record.toBuilder().stream(appendValue).build().toBytes();
+        }
+        return TableSegmentEntry.versioned(key.getBytes(StandardCharsets.UTF_8), updatedBytes,
+                                           entry.getKey().getVersion().getSegmentVersion());
+    }
+
+    /**
+     * Remove Tags for the specified Stream.
+     * @param stream Stream for which tags are removed.
+     * @param tags Stream tags.
+     * @param context Context.
+     * @return A future which completes once the update is complete.
+     */
+    public CompletableFuture<Void> removeTagsUnderScope(String stream, Set<String> tags, OperationContext context) {
+        return getAllStreamTagsInScopeTableNames(stream, context)
+                .thenCompose(table -> Futures.allOf(tags.stream()
+                                                        .parallel()
+                                                        .map(key -> storeHelper.getAndUpdateValues(table, key,
+                                                                                                   e -> removeStreamFromEntry(stream, e),
+                                                                                                   this::isEmptyTagRecord,
+                                                                                                   context.getRequestId()))
+                                                        .collect(Collectors.toList())));
+    }
+
+    private TableSegmentEntry removeStreamFromEntry(String removeValue, TableSegmentEntry currentEntry) {
+        String k = currentEntry.getKey().getKey().toString(StandardCharsets.UTF_8);
+        byte[] array = storeHelper.getArray(currentEntry.getValue());
+        byte[] updatedBytes;
+        if (array.length == 0) {
+            updatedBytes = TagRecord.builder().tagName(k).build().toBytes();
```

Comment:
nit: we can simply return current entry from here without having to create a new object.

Reply:
yes, improved the code.